File: go/min-heap/heap.go
package main

import "fmt"

type MinHeap struct {
	array []int
	size  int
}

func NewMinHeap(capacity int) *MinHeap {
	return &MinHeap{
		array: make([]int, 0, capacity),
		size:  0,
	}
}

func (h *MinHeap) insert(value int) {
	h.array = append(h.array, value)
	h.size++
	h.heapifyUp(h.size - 1)
}

func (h *MinHeap) extractMin() int {
	if h.size == 0 {
		fmt.Println("Heap is empty")
		return -1
	}
	min := h.array[0]
	h.array[0] = h.array[h.size-1]
	h.array = h.array[:h.size-1]
	h.size--
	h.heapifyDown(0)
	return min
}

func (h *MinHeap) getMin() int {
	if h.size == 0 {
		fmt.Println("Heap is empty")
		return -1
	}
	return h.array[0]
}

func (h *MinHeap) heapifyUp(index int) {
	for index > 0 && h.array[(index-1)/2] > h.array[index] {
		h.swap((index-1)/2, index)
		index = (index - 1) / 2
	}
}

func (h *MinHeap) heapifyDown(index int) {
	smallest := index
	left := 2*index + 1
	right := 2*index + 2

	if left < h.size && h.array[left] < h.array[smallest] {
		smallest = left
	}

	if right < h.size && h.array[right] < h.array[smallest] {
		smallest = right
	}

	if smallest != index {
		h.swap(index, smallest)
		h.heapifyDown(smallest)
	}
}

func (h *MinHeap) swap(i, j int) {
	h.array[i], h.array[j] = h.array[j], h.array[i]
}

func (h *MinHeap) printHeap() {
	fmt.Print("Min Heap: ")
	for i := 0; i < h.size; i++ {
		fmt.Printf("%d ", h.array[i])
	}
	fmt.Println()
}

File: zig/min-heap/src/heap.zig
const std = @import("std");

pub const MinHeap = struct {
    heap: []i32,
    size: usize,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, capacity: usize) !MinHeap {
        const heap = try allocator.alloc(i32, capacity);
        return MinHeap{
            .heap = heap,
            .size = 0,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *MinHeap) void {
        self.allocator.free(self.heap);
    }

    pub fn insert(self: *MinHeap, value: i32) !void {
        if (self.size >= self.heap.len) {
            std.debug.print("Heap is full\n", .{});
            return;
        }

        self.heap[self.size] = value;
        self.heapifyUp(self.size);
        self.size += 1;
    }

    pub fn extractMin(self: *MinHeap) ?i32 {
        if (self.size == 0) {
            std.debug.print("Heap is empty\n", .{});
            return null;
        }

        const min = self.heap[0];
        self.size -= 1;
        if (self.size > 0) {
            self.heap[0] = self.heap[self.size];
            self.heapifyDown(0);
        }
        return min;
    }

    pub fn getMin(self: MinHeap) ?i32 {
        if (self.size == 0) {
            std.debug.print("Heap is empty\n", .{});
            return null;
        }
        return self.heap[0];
    }

    fn heapifyUp(self: *MinHeap, index: usize) void {
        var current = index;
        while (current > 0) {
            const parent = (current - 1) / 2;
            if (self.heap[current] < self.heap[parent]) {
                self.swap(current, parent);
                current = parent;
            } else break;
        }
    }

    fn heapifyDown(self: *MinHeap, index: usize) void {
        var smallest = index;
        const left = 2 * index + 1;
        const right = 2 * index + 2;

        if (left < self.size and self.heap[left] < self.heap[smallest]) {
            smallest = left;
        }

        if (right < self.size and self.heap[right] < self.heap[smallest]) {
            smallest = right;
        }

        if (smallest != index) {
            self.swap(index, smallest);
            self.heapifyDown(smallest);
        }
    }

    fn swap(self: *MinHeap, i: usize, j: usize) void {
        const temp = self.heap[i];
        self.heap[i] = self.heap[j];
        self.heap[j] = temp;
    }

    pub fn printHeap(self: MinHeap) void {
        std.debug.print("Min Heap: ", .{});
        var i: usize = 0;
        while (i < self.size) : (i += 1) {
            std.debug.print("{} ", .{self.heap[i]});
        }
        std.debug.print("\n", .{});
    }
};
